tests/lang_types/typeclass/tbuiltin_type_classes_bind_once.nim

```diff
@@ -0,0 +1,55 @@
+discard """
+  description: '''
+    All built-in type classes are *named*, and they thus have bind-once
+    behaviour. ``typedesc`` is an exception.
```

Comment:
IMO, the manual's claim of bind once seems like an unprincipled decision and instead one of convenience wrt implementation bugs at the time. I'm not firm on this, but I can't think of a good reason to have it work this way.

I don't think any adjustments need to be made per my comment above, I'd prefer to discuss it further before settling on something.

Briefly, I think the following are saying different things:

```nim
# type classes
proc foo(a: array, b: array) # two params, each of some type satisfying `array`
proc foo(a, b: array) # same as above, as I believe pragmas require a copy of the type ast during normalization

# array[T]/array[T:any]
proc foo[T](a: array[T], b: array[T]) # only one specific `array[T]`
proc foo[T](a, b: array[T]) # same as above, even though `array[T]` is copied per param during normalization due to pragmas, `T` only binds once
proc foo[T: any](a: array[T], b: array[T]) # same as above, this is technically the most "expanded" version of the three
```

Again, I don't think this PR should necessarily change its direction.

Reply:
Once case to consider could be the following:
```nim
proc p(a: array): array # the return type is the same as the parameter's type
```
I'm not entirely sure, but this could be useful.

Anyhow, I don't think I'd object to removing the "bind once" behaviour. For now, my goal was to make the language work as specified in the manual.

Reply:
> Once case to consider could be the following:
> 
> ```nim
> proc p(a: array): array # the return type is the same as the parameter's type
> ```
> 
> I'm not entirely sure, but this could be useful.

Yup, it's a bit convenient but it might be better if the return type was `typeOf(x)`. Type classes and implicit generics are much weirder the more I think about them.

> Anyhow, I don't think I'd object to removing the "bind once" behaviour. For now, my goal was to make the language work as specified in the manual.

That makes sense, it's also why I don't think this pr should change because of it.